src/Framework/LazyFormattedBuildEventArgs.cs

```diff
@@ -125,37 +127,26 @@ public override string Message
         /// <param name="writer">Binary writer which is attached to the stream the event will be serialized into.</param>
         internal override void WriteToStream(BinaryWriter writer)
         {
-            // Locking is needed here as this is invoked on the serialization thread,
-            // whereas a local logger (a distributed logger) may concurrently invoke this.Message
-            // which will trigger formatting and thus the exception below
-            lock (locker)
+            object argsOrMessage = argumentsOrFormattedMessage;
+            if (argsOrMessage is object[] arguments && arguments.Length > 0)
             {
-                bool hasArguments = arguments != null;
-                base.WriteToStream(writer);
+                base.WriteToStreamWithExplicitMessage(writer, base.Message);
+                writer.Write(arguments.Length);
 
-                if (hasArguments && arguments == null)
+                foreach (object argument in arguments)
                 {
-                    throw new InvalidOperationException("BuildEventArgs has formatted message while serializing!");
+                    // Arguments may be ints, etc, so explicitly convert
+                    // Convert.ToString returns String.Empty when it cannot convert, rather than throwing
+                    writer.Write(Convert.ToString(argument, CultureInfo.CurrentCulture));
```

Comment:
It just occurs to me that getting CurrentCulture might not be super cheap, should we do it outside the loop and cache in a local variable?